### ISSUE TITLE
[java] MethodReturnsInternalArray does not consider static final fields and fields initialized with empty array

### ISSUE THREAD

**Issue by oowekyala** [pmd/pmd#3160]
**Affects PMD Version:** tbd

**Rule:** MethodReturnsInternalArray, https://pmd.github.io/latest/pmd_rules_java_bestpractices.html#methodreturnsinternalarray

**Description:** Static final fields are currently ignored as per https://sourceforge.net/p/pmd/bugs/1475/. However, this is only safe because the array has length zero. Sharing non-empty arrays is problematic whether the field is final or not, so it doesn't make sense to forbid one and allow the other.

**Code Sample demonstrating the issue:**

```java
public class MyClass {
    private static final String[] FOO_BAR = new String[] { "foo", "bar" };
    public String[] call() { return FOO_BAR; }
}
```

Second example:
```java
public class MyClass {
    private String[] fooBar = new String[0];
    public String[] call() { return fooBar; }
}
```

**Expected outcome:** A violation should be reported

**Comment by oowekyala** [comment]
So actually master is not affected, this is just a regression in #3127 - and there's no test on master for this.

**Comment by adangel** [comment]
I'm reopening, because on pmd7 we now actually fixed a false-negative: If a non-final field was initialized with an empty array, we ignored it. That's why the violations on [SQLErrorCodes.java](https://github.com/spring-projects/spring-framework/blob/v5.0.6.RELEASE/spring-jdbc/src/main/java/org/springframework/jdbc/support/SQLErrorCodes.java) are new: the fields look like this:

```java
private String[] badSqlGrammarCodes = new String[0];
```

Since the field is not final, it can be reassigned, hence it could also be a non-empty array.

The fix is trivial ([in this loop](https://github.com/pmd/pmd/blob/d8ddb09b35b49829c85a9d0b70bfcfc70cd62884/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/bestpractices/MethodReturnsInternalArrayRule.java#L117) we should only consider final fields) - I'll fix it for pmd6 and merge then into pmd7.
